src/Filters/remez_fir.jl

```diff
@@ -0,0 +1,829 @@
+ *  MAGNITUDE.  NOTE THAT THE PARAMETER FREQ IS THE
+ *  VALUE OF NORMALIZED FREQUENCY NEEDED FOR EVALUATION.
+ *-----------------------------------------------------------------------
+ */
+"""
+function eff(freq::Float64, fx::AbstractVector, lband::Integer, filter_type::RemezFilterType)
+    if filter_type == filter_type_differentiator
+        return fx[lband] * freq
+    else 
+        return fx[lband]
+    end
+end
+
+"""
+wate(freq::Float64, fx::AbstractVector, wtx::AbstractVector, lband::Integer, filter_type::RemezFilterType)
+/*
+ *-----------------------------------------------------------------------
+ * FUNCTION: wate
+ *  FUNCTION TO CALCULATE THE WEIGHT FUNCTION AS A FUNCTION
+ *  OF FREQUENCY.  SIMILAR TO THE FUNCTION eff, THIS FUNCTION CAN
+ *  BE REPLACED BY A USER-WRITTEN ROUTINE TO CALCULATE ANY
+ *  DESIRED WEIGHTING FUNCTION.
+ *-----------------------------------------------------------------------
+ */
+"""
+function wate(freq::Float64, fx::AbstractVector, wtx::AbstractVector, lband::Integer, filter_type::RemezFilterType)
+    if filter_type != filter_type_differentiator
+        return wtx[lband]
+    end
+    if fx[lband] >= 0.0001
+        return wtx[lband] / freq
+    end
+    wtx[lband]
+end
+
+
+"""
+build_grid(numtaps, bands, desired, weight, grid_density)
+return "grid" and "des" and "wt" arrays
+"""
+function build_grid(numtaps, bands, desired, weight, grid_density, filter_type::RemezFilterType)
+    # translate from scipy remez argument names
+    L = numtaps
+    M = L ÷ 2    # integer divide (truncated)
+    grid_spacing = 0.5 / (grid_density * M)  # "delf" or "delta f"
+
+    lgrid = grid_density
+    dimsize = Int( ceil(numtaps/2.0 + 2) )
+    wrksize = grid_density * dimsize
+
+    grid = zeros(Float64, wrksize)  # the array of frequencies, between 0 and 0.5
+    des = zeros(Float64, wrksize)   # the desired function on the grid
+    wt = zeros(Float64, wrksize)    # array of weights
+
+    fx = desired
+    wtx = weight
+    
+    nbands = length(desired)
+    edge = bands
+    nfilt = numtaps
+
+    neg = filter_type != filter_type_bandpass
+    nodd = isodd(nfilt)
+    nfcns = nfilt ÷ 2
+    if nodd && !neg
+        nfcns = nfcns + 1
+    end
+
+    #
+    # SET UP THE DENSE GRID. THE NUMBER OF POINTS IN THE GRID
+    # IS (FILTER LENGTH + 1)*GRID DENSITY/2
+    #
+    grid[1] = edge[1]
+    delf = lgrid * nfcns
+    delf = 0.5 / delf
+    if neg
+        if edge[1] < delf
+            grid[1] = delf
+        end
+    end
+    j = 1
+    l = 1
+    lband = 1
+
+    #
+    # CALCULATE THE DESIRED MAGNITUDE RESPONSE AND THE WEIGHT
+    # FUNCTION ON THE GRID
+    #
+    while true
+        fup = edge[l + 1]
+        while true
+            temp = grid[j]
+            des[j] = eff(temp,fx,lband,filter_type)
+            wt[j] = wate(temp,fx,wtx,lband,filter_type)
+            j += 1
+            if j > wrksize
+                # too many points, or too dense grid
+                return -1
```

Comment:
If this is ever hit, it will give a rather obscure error message because this case is not covered in `remez`. Should probably throw an error from here, though I'm not sure under what circumstances this can happen.